app/gui2/src/stores/imports.ts

```diff
@@ -0,0 +1,589 @@
+  const from = ast.tryMap((import_) => import_.from?.body)
+  const as = ast.tryMap((import_) => import_.as?.body)
+  const import_ = ast.tryMap((import_) => import_.import.body)
+  const all = ast.tryMap((import_) => import_.all)
+  const hiding = ast.tryMap((import_) => import_.hiding?.body)
+  const module =
+    from != null ? parseQualifiedName(from) : import_ != null ? parseQualifiedName(import_) : null
+  if (!module) return null
+  if (all) {
+    const except = (hiding != null ? parseIdents(hiding) : []) ?? []
+    return {
+      from: module,
+      imported: { kind: 'All', except },
+    }
+  } else if (from && import_) {
+    const names = parseIdents(import_) ?? []
+    return {
+      from: module,
+      imported: { kind: 'List', names },
+    }
+  } else if (import_) {
+    const alias = as ? parseIdent(as) : null
+    return {
+      from: module,
+      imported: alias ? { kind: 'Module', alias } : { kind: 'Module' },
+    }
+  } else {
+    console.error('Unrecognized import', ast.debug())
+    return null
+  }
+}
+
+export type ModuleName = QualifiedName
+
+/** Information about parsed import statement. */
+export interface Import {
+  from: ModuleName
+  imported: ImportedNames
+}
+
+export type ImportedNames = Module | List | All
+
+/** import Module.Path (as Alias)? */
+export interface Module {
+  kind: 'Module'
+  alias?: Identifier
+}
+
+/** from Module.Path import (Ident),+ */
+export interface List {
+  kind: 'List'
+  names: Identifier[]
+}
+
+/** from Module.Path import all (hiding (Ident),*)? */
+export interface All {
+  kind: 'All'
+  except: Identifier[]
+}
+
+// ========================
+// === Required imports ===
+// ========================
+
+/** Import required for the suggestion entry. */
+export type RequiredImport = QualifiedImport | UnqualifiedImport
+
+/** import Module.Path */
+export interface QualifiedImport {
+  kind: 'Qualified'
+  module: QualifiedName
+}
+
+/** from Module.Path import SomeIdentifier */
+export interface UnqualifiedImport {
+  kind: 'Unqualified'
+  from: QualifiedName
+  import: Identifier
+}
+
+/** Get the string representation of the required import statement. */
+export function requiredImportToText(value: RequiredImport): string {
+  switch (value.kind) {
+    case 'Qualified':
+      return `import ${value.module}`
+    case 'Unqualified':
+      return `from ${value.from} import ${value.import}`
+  }
+}
+
+/** A list of required imports for specific suggestion entry */
+export function requiredImports(db: SuggestionDb, entry: SuggestionEntry): RequiredImport[] {
+  switch (entry.kind) {
+    case SuggestionKind.Module:
+      return entry.reexportedIn
+        ? [
+            {
+              kind: 'Unqualified',
+              from: entry.reexportedIn,
+              import: entry.name,
+            },
+          ]
+        : [
+            {
+              kind: 'Qualified',
+              module: entryQn(entry),
```

Comment:
`entry.definedIn` should work here, and has better semantics IMO.

Reply:
Done